libr/anal/type.c

```diff
@@ -120,20 +120,258 @@ R_API void r_anal_save_parsed_type(RAnal *anal, const char *parsed) {
 }
 
 static int typecmp(const void *a, const void *b) {
-	char *type1 = (char *) a;
-	char *type2 = (char *) b;
-	return strcmp (type1 , type2);
+	return strcmp (a, b);
 }
 
 R_API RList *r_anal_types_from_fcn(RAnal *anal, RAnalFunction *fcn) {
 	RListIter *iter;
 	RAnalVar *var;
 	RList *list = r_anal_var_all_list (anal, fcn);
-        RList *type_used = r_list_new ();
-        r_list_foreach (list, iter, var) {
-		r_list_append (type_used , var->type);
-        }
-	RList *uniq = r_list_uniq (type_used , typecmp);
+	RList *type_used = r_list_new ();
+	r_list_foreach (list, iter, var) {
+		r_list_append (type_used, var->type);
+	}
+	RList *uniq = r_list_uniq (type_used, typecmp);
 	r_list_free (type_used);
 	return uniq;
 }
+
+static void enum_type_fini(void *e, void *user) {
+	(void)user;
+	RAnalEnumCase *cas = e;
+	free ((char *)cas->name);
+}
+
+static void struct_type_fini(void *e, void *user) {
+	(void)user;
+	RAnalStructMember *member = e;
+	free ((char *)member->name);
+	free ((char *)member->type);
+}
+
+static void union_type_fini(void *e, void *user) {
+	(void)user;
+	RAnalUnionMember *member = e;
+	free ((char *)member->name);
+	free ((char *)member->type);
+}
+
+static RAnalBaseType *get_enum_type(RAnal *anal, const char *sanitized_name) {
+	r_return_val_if_fail (anal && sanitized_name, NULL);
+
+	RAnalBaseType *base_type = R_NEW0 (RAnalBaseType);
+	if (!base_type) {
+		return NULL;
+	}
+	base_type->kind = R_ANAL_BASE_TYPE_KIND_ENUM;
+
+	RAnalBaseTypeEnum base_enum;
+	char *key = r_str_newf ("enum.%s", sanitized_name);
+	if (!key) {
+		free (base_type);
+		return NULL;
+	}
+	char *members = sdb_get (anal->sdb_types, key, NULL);
+	free (key);
+
+	RVector cases;
+	r_vector_init (&cases, sizeof (RAnalEnumCase), enum_type_fini, NULL);
+
+	if (!r_vector_reserve (&cases, (size_t)sdb_alen (members))) {
+		goto error;
+	}
+
+	char *cur;
+	sdb_aforeach (cur, members) {
+		char *val_key = r_str_newf ("enum.%s.%s", sanitized_name, cur);
+		if (!val_key) {
+			goto error;
+		}
+		const char *value = sdb_const_get (anal->sdb_types, val_key, NULL);
+		free (val_key);
+
+		RAnalEnumCase cas = { .name = strdup (cur), .val = strtol (value, NULL, 16) };
+
+		void *element = r_vector_push (&cases, &cas); // returns null if  no space available
+		if (!element) {
+			goto error;
+		}
+
+		sdb_aforeach_next (cur);
+	}
+	free (members);
+	base_enum.cases = cases;
+	base_type->enum_data = base_enum;
+
+	return base_type;
+
+error:
+	free (base_type);
+	free (members);
+	r_vector_fini (&cases);
+	return NULL;
+}
+
+static RAnalBaseType *get_struct_type(RAnal *anal, const char *sanitized_name) {
```

Comment:
see? santized name is used here, below its using name_sanitized. better be consistent and follow the same naming across the code